Implement structure encoding

mojoX911 opened the pull request:
This PR implements our own Encoding/Decoding using `bitcoin::consensus::encode::Encodable` traits. 

I decided to go with our custom traits because that gives us flexibility and we need encoding for stuffs that isn't covered by rust-bitcoin (like `Signature`, `Publickey` etc).

The commits look bulky, but its mostly straight forward change.

Related #19 

There is a typo in function name `serialisable`, this is intentional to avoid conflict with `serde::serializable` and will be fixed once we move away from `serde` completely.

GeneFerneau commented on src/serialization.rs:
```diff
@@ -0,0 +1,171 @@
+// The underlying encoding of primitve datatyes and structures are done
+// via bitcoin::consensus::encode::Encodale trait.
+pub trait Serializable {
+    fn serialise<W: std::io::Write>(&self, w: W) -> Result<usize, SerialisationError>;
```
Maybe instead of the intentional typo here, below, and in the error type, you can use the normal spelling. When there is a naming import conflict with `serde::Serializable` you can import the local function as `TeleportSerializable` or keep the mod path like `serialization::Serializable`.

Small typos like this, even if intentional, make the code harder to read, IMHO. The extra mental load of "does this come from our local serialization code, and I need to type serialise" also seems like it would get annoying pretty quickly.

GeneFerneau commented on src/serialization.rs:
```diff
@@ -0,0 +1,171 @@
+    }
+}
+
+impl Serializable for SecretKey {
```
[rust-secp256k1](https://github.com/rust-bitcoin/rust-secp256k1/blob/master/src/key.rs#L218-L242) has a `serde` implementation for `SecretKey`

GeneFerneau commented on src/serialization.rs:
```diff
@@ -0,0 +1,171 @@
+    }
+}
+
+impl Serializable for PublicKey {
+    fn serialise<W: std::io::Write>(&self, mut w: W) -> Result<usize, SerialisationError> {
+        Ok(w.write(&self.to_bytes()[..])?)
+    }
+}
+
+impl Deserilizable for PublicKey {
+    fn deserialise<R: std::io::Read>(mut r: R) -> Result<Self, SerialisationError> {
+        let mut bytes = [0u8; 33];
+        r.read_slice(&mut bytes)?;
+        Ok(PublicKey::from_slice(&bytes)?)
+    }
+}
```
[rust-secp256k1](https://github.com/rust-bitcoin/rust-secp256k1/blob/master/src/key.rs#L444-L468) has a `serde` implementation for `PublicKey`

GeneFerneau commented on src/serialization.rs:
```diff
@@ -0,0 +1,171 @@
+    }
+}
+
+impl Serializable for SecretKey {
+    fn serialise<W: std::io::Write>(&self, mut w: W) -> Result<usize, SerialisationError> {
+        Ok(w.write(&self[..])?)
+    }
+}
+
+impl Deserilizable for SecretKey {
+    fn deserialise<R: std::io::Read>(mut r: R) -> Result<Self, SerialisationError> {
+        let mut secret_key_bytes = [0u8; 32];
+        r.read_slice(&mut secret_key_bytes)?;
+        Ok(SecretKey::from_slice(&secret_key_bytes)?)
+    }
+}
```
[rust-secp256k1](https://github.com/rust-bitcoin/rust-secp256k1/blob/master/src/key.rs#L218-L242) has a `serde` implementation for `SecretKey`

GeneFerneau commented on src/serialization.rs:
```diff
@@ -0,0 +1,171 @@
+    }
+}
+
+impl Serializable for Script {
+    fn serialise<W: std::io::Write>(&self, mut w: W) -> Result<usize, SerialisationError> {
+        Ok(self.consensus_encode(&mut w)?)
+    }
+}
+
+impl Deserilizable for Script {
+    fn deserialise<R: std::io::Read>(mut r: R) -> Result<Self, SerialisationError> {
+        Ok(Self::consensus_decode(&mut r)?)
+    }
+}
```
[rust-bitcoin](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/src/blockdata/script.rs#L824-L894) has a `serde` and `consensus_en/decode` implementation for `Script`

GeneFerneau commented on src/serialization.rs:
```diff
@@ -0,0 +1,171 @@
+    }
+}
+
+impl Serializable for Signature {
+    fn serialise<W: std::io::Write>(&self, mut w: W) -> Result<usize, SerialisationError> {
+        Ok(w.write(&self.serialize_compact()[..])?)
+    }
+}
+
+impl Deserilizable for Signature {
+    fn deserialise<R: std::io::Read>(mut r: R) -> Result<Self, SerialisationError> {
+        let mut sig_buff = [0u8; 64];
+        r.read_exact(&mut sig_buff)?;
+        Ok(Signature::from_compact(&sig_buff[..])?)
+    }
+}
```
[rust-secp256k1](https://github.com/rust-bitcoin/rust-secp256k1/blob/master/src/lib.rs#L434-L458) has a `serde` implementation for `Signature`

GeneFerneau commented:
> There is a typo in function name serialisable, this is intentional to avoid conflict with serde::serializable and will be fixed once we move away from serde completely.

What is the purpose of moving away from `serde`? It's a pretty common crate for serialization across not only `rust-bitcoin` crates, but the entire Rust ecosystem. What gains do you see in moving away from it?

> I decided to go with our custom traits because that gives us flexibility and we need encoding for stuffs that isn't covered by rust-bitcoin (like Signature, Publickey etc).

`rust-bitcoin` and `rust-secp256k1` have serialization impls for nearly everything you have in `src/serialization.rs`. Are the impls just convenience tools over the implementations provided upstream?

What extra flexibility is provided by the custom impls?

mojoX911 commented:
Thanks @GeneFerneau for the review. 

As far as my understanding `serde` implementations in `rust-bitcoin` are done for human readable representation of structure data, like `json` or `yaml`. This is useful when writing wallets that displays and store data for user. It also requires the user of the lib to implement their own `Serializer` or use existing serializer like `serde_json`.

While the consensus en/decode are provided to produce byte array representation of structures following bitcoin's consensus. These are used mostly in network communication. Which is what we are aiming for here.

We are already using serde for everything. We don't need `rust-bitcoin`'s serde to do it again.  The intent of this PR is to provide byte encoding of message structure following bitcoin's consensus (there might be other more efficient byte encoding out there, but mostly I have seen other projects using bitcoin's encoding).

The reason we want a separate implementation is because bitcoin's encoding doesn't provide all the primitives. Probably because they are not much used in bitcoin network messages. But we need those primitive in our messages, and we also need to extend them to derived structures (like `Vec<Signature>`). Rust won't allow us to implementation foreign traits on foreign types, thus the new implementation.

regarding the name I don't think there is any need to have serde serialization derived for message structures if we are going for bitcoin's encoding. So we should be able to use "serialize" for everything eventually.

The entire purpose of having a custom byte serialization is to have only one `ser/deserialize` function that covers everything in the library. Otherwise you have to keep track on which structures are serializable by with method, because rust wont allow using foreign traits flexibily, for dep consistency reasons. 

So the wrong name is transitory, if its bothering too much we can use explicit declaration everywhere to remove the conflict.

GeneFerneau commented:
> The entire purpose of having a custom byte serialization is to have only one ser/deserialize function that covers everything in the library.

That makes sense. Now I understand better having local trait(s) to cover everything in the crate, instead of using the stuff already present in `rust-bitcoin` and `rust-secp256k1`.

> So the wrong name is transitory, if its bothering too much we can use explicit declaration everywhere to remove the conflict.

If changing the way everything in the crate is serialized/encoded, doing it in one go makes more sense to me than using an intentionally misspelled function name. "The most permanent things are temporary solutions..." - someone on the internet probably

I don't really understand your reasoning on the misspelled function, if using local traits for serialization, there should be no conflicts. Am I missing something?

> The intent of this PR is to provide byte encoding of message structure following bitcoin's consensus

This seems reasonable for stuff that needs to follow consensus (inclusion in blocks), but for communication rounds something like [CBOR](https://github.com/bitcoin-teleport/teleport-transactions/issues/19#issuecomment-893873307) probably fits better.

To distinguish between the two, you could add a `net_serialize` function to `Serializable` for over-the-wire encoding, and use `serialize` to mean "consensus encode". What do you think?

GeneFerneau commented:
> We will be sending around tx and other data through our connections, and if we for some reason need to broadcast a transaction in the bitcoin p2p network, we need it to be serialized as per "bitcoin consensus". That's the job of consensus_encode().

I understand that, and that is what I meant: use `net_serialize` for places where `consensus_encode` needs to be called.

> Feel free to suggest one and I will update the PR.

Since we're not importing upstream implementations of `serde::Serialize`, why can the local impls you have in this PR just use that trait? Then, for the stuff that needs to use `consensus_encode`, something like:

```rust
pub trait NetSerialize {
    fn net_serialize(...) {
        // ...
        consensus_encode(...);
    }

    fn net_deserialize(...) {
        // ...
        consensus_decode(...);
    }
}
```

mojoX911 commented:
> Since we're not importing upstream implementations of serde::Serialize,

We are using `serde:Serlialize` here 
https://github.com/bitcoin-teleport/teleport-transactions/blob/b67dab5ce1ae04e4de7665004d1ffba944ef2318/src/messages.rs#L18

Which I didn't remove in  this PR, would be done on a subsequent one if required. This causes the name conflict. 

Ya `net-serialize` sounds good to me. Would update with that.

GeneFerneau commented:
> Which I didn't remove in this PR, would be done on a subsequent one if required. This causes the name conflict.

I would recommend doing your custom `serde` impl, and removing the automatic derive that causes the conflict.

GeneFerneau commented:
Overall, changes look great! I've done another quick review on [b05b838](https://github.com/bitcoin-teleport/teleport-transactions/pull/31/commits/b05b83802adbcb068ebb43c7a6e35c44ace2b2c1).

A couple nits, but the code looks really solid overall.

mojoX911 commented:
> > Which I didn't remove in this PR, would be done on a subsequent one if required. This causes the name conflict.
> 
> I would recommend doing your custom `serde` impl, and removing the automatic derive that causes the conflict. Unless the `NetSerialize` impls take care of all the conflicts.

Ya if we just remove the derive macros, The name conflicts will be gone. But then we would also need to replace all serde calls inside the protocol. I plan on doing that after this PR is merged, subjected to @chris-belcher 's opinion. 

Could have done it here also, but that would create a change set I feel too big for a single PR.

mojoX911 commented:
I am also wondering if its better to have `net-serialization` impls in the `serialization.rs` than in the `messgae.rs` itself, from code oragnisation perspective. 

Is it better to put encoding impls in a single module or in each structures own module? @GeneFerneau @chris-belcher let me know if you have any thought/prefs.

I have seen both of them used in different projects.

GeneFerneau commented:
> Is it better to put encoding impls in a single module or in each structures own module?

IMHO, having all serialization stuff in `serialization.rs` makes sense. Of course, will defer to @chris-belcher on this, though.

GeneFerneau commented on src/serialization.rs:
```diff
@@ -0,0 +1,171 @@
+pub const MAX_VEC_SIZE: usize = 4_000_000;
+
+#[derive(Debug)]
+pub enum NetSerialisationError {
```
nit:

```suggestion
pub enum NetSerializationError {
```

GeneFerneau commented on src/serialization.rs:
```diff
@@ -0,0 +1,171 @@
+}
+
+// Helper function to quickly generate serialized data
+pub fn serialise<T: NetSerialize>(thing: &T) -> Result<Vec<u8>, NetSerialisationError> {
```
nit:

```suggestion
pub fn serialize<T: NetSerialize>(thing: &T) -> Result<Vec<u8>, NetSerialisationError> {
```

GeneFerneau commented on src/serialization.rs:
```diff
@@ -0,0 +1,171 @@
+// The underlying encoding of primitve datatyes and structures are done
+// via bitcoin::consensus::encode::Encodale trait.
+pub trait NetSerialize {
+    fn net_serialise<W: std::io::Write>(&self, w: W) -> Result<usize, NetSerialisationError>;
```
nit:

```suggestion
    fn net_serialize<W: std::io::Write>(&self, w: W) -> Result<usize, NetSerialisationError>;
```

chris-belcher commented on src/serialization.rs:
```diff
@@ -0,0 +1,171 @@
+use bitcoin::{PublicKey, Script, VarInt};
+
+// Maximum size, in bytes, of a vector we are allowed to decode
+// TODO: Decide weather its useful
```
nit: typo "whether"

chris-belcher commented:
Not sure about the concept of the PR now that I just read about [serde_cbor](https://github.com/bitcoin-teleport/teleport-transactions/issues/19#issuecomment-893873307). Could we use that instead? What are the upsides/downsides

mojoX911 commented:
> Not sure about the concept of the PR now that I just read about serde_cbor. Could we use that instead? What are the upsides/downsides

We can. Then we have to consider the following situations
 - Our encoding will diverge from bitcoin's network encodings. If we decide to send our transactions and other data to some other apps in future, there would be interoperability issues. They all have to understand cbor. In practice I have encountered most apps use bitcoin's network encdoing. That's kinda the defacto standard.
 - We will need to implement cbor for all bitcoin primitive types. Here we get most of stuffs for free.
 
In principle they are both similar and solves our purpose. My only worry is interoperability in future.

mojoX911 commented:
Updated with nit comments.

GeneFerneau commented:
> In principle they are both similar and solves our purpose. My only worry is interoperability in future.

Both points you raise make sense to me. The main reason I brought up CBOR was for use-cases where JSON was being used for serialization, not consensus encoding. I agree for consensus encoding uses, it is more reasonable to use the de-facto standard (`consensus_{de,en}code`).

If there are other cases where JSON serialization is used, I would advocate for CBOR, especially if the bytes are going over the network.

chris-belcher commented:
> Our encoding will diverge from bitcoin's network encodings. If we decide to send our transactions and other data to some other apps in future, there would be interoperability issues

JoinMarket has the same taker/maker setup and in its several years of existence there's never been a need for other apps to implement the taker/maker protocol. I don't think we'll ever end up in a situation where other apps are talking the teleport protocol between makers and takers. Interoperability is not an issue at all since it's overwhelmingly likely that this app will only ever talk to other instances of itself.

> We will need to implement cbor for all bitcoin primitive types. Here we get most of stuffs for free.

Is it possible to write something like `#[derive(Debug, Serialize, Deserialize)]` which does it all automatically? That's how it's done with serde.

GeneFerneau commented:
> I am not too familiar with how serde works, but my guess is we would need to have some serde::cbor impls for primitive bitcoin types.

AFAICT, the `Serialize/Deserialize` traits are general across all `serde` implementations. So when you implement actually reading/writing values, you would use something like `serde_cbor::to_vec(&some_struct_that_impls_Serialize)`, instead of the `serde_json` equivalent. Similar for reading values, you would use `serde_cbor::Value`.

mojoX911 commented:
I have been playing around with cbor (both serde and other alternatives). So far they can't seem to handle our object serializations, and I am not sure why.

Here's a demonstration
```
    #[test]
    fn test2() {
        let expected_pubkey = PublicKey::from_str(
            "03bf98c86c3d536136378cf43ac42861ece609de87f5a44e19b730e8e9bd791938",
        )
        .unwrap();

        let x = MakerToTakerMessage::Offer(Offer {
            absolute_fee: 1000,
            amount_relative_fee: 0.005,
            max_size: 200,
            min_size: 10000,
            tweakable_point: expected_pubkey,
        });

        let mut ciborum_encoded = Vec::new();
        ciborium::ser::into_writer(&x, &mut ciborum_encoded).unwrap();

        let serde_encoded = serde_cbor::to_vec(&x).unwrap();

        let message1 : MakerToTakerMessage = ciborium::de::from_reader(&ciborum_encoded[..]).unwrap();
        let message2 : MakerToTakerMessage = serde_cbor::from_slice(&serde_encoded).unwrap();
    }
```

I am using two different impls of cbor (`serde_cbor` and `ciborium`). In both the cases the last two unwrap will result in errors.

GeneFerneau commented:
> I am using two different impls of cbor (serde_cbor and ciborium). In both the cases the last two unwrap will result in errors.

Just reproduced the errors locally, not sure the best thing to do about it. Wasn't aware about `serde_cbor` locking their repo.

IMHO, think the best thing to do is keep the `consensus_[de,en]code` methods for now, and revisit the CBOR stuff when the libraries are more mature.

I can dig into the errors as time permits, since it seems like an internal lib error (more likely), or we are using the lib wrong (less likely). Thanks for trying to get this working.

mojoX911 commented:
Yeah makes sense.. It seems CBOR deps aren't mature enough to handle our data, especially large byte numbers. 

I will finalize the hand encoding way.. For now that seems like the only option.

mojoX911 commented:
I am not sure, but one thing I am expecting to face while integrating byte encoding into the networking modules is how to separate messages without deliminator. For json we are using `\n` which works fine. But this will not work for byte encoding, as `\n`'s byte value can occur within the message data. 

Any suggestion or refs on how to handle such situation would be helpful. Thanks..

GeneFerneau commented:
> Any suggestion or refs on how to handle such situation would be helpful. Thanks..

Like mentioned in the StackOverflow post you linked, prefix-length encoding is pretty useful for network encoding. A number of places in Bitcoin use minimally encoded integers for the length value, followed by that length of data. If I'm not wrong, this is what `consensus_encode` is doing.

mojoX911 commented:
@GeneFerneau yes right. That seems like the only way encoding stuffs via binary. But given our message data can be of arbitrary size, it seems things gonna get complex pretty fast.

I am not exactly sure weather the complexity tradeoff of doing and maintaining it, is less than the data overhead tradeoff of `json` encoding.

Because of easy delimination, the read buffer logic is also simple now. Just read until next line..

mojoX911 commented:
Thanks @GeneFerneau for the suggestion. I tried doing something like that. Have a `VarInt` in front of the message indicating how long it is. Then fill a buffer of that specific size and then try to decode the message. 

For that I modified the `send` and `read` functions as below
```rust
async fn send_message(
    socket_writer: &mut WriteHalf<'_>,
    message: &MakerToTakerMessage,
) -> Result<(), Error> {
    let mut message_bytes = vec![];
    
    let len = message.net_serialize(&mut message_bytes)?;
    let var_len = VarInt(len as u64);
    
    let mut result = vec![];
    var_len.consensus_encode(&mut result).map_err(|e| Error::Serialisation(e.into()))?;
    
    result.extend_from_slice(&mut message_bytes);

    socket_writer.write_all(&mut result).await?;
    Ok(())
}

async fn read_message(reader: &mut BufReader<ReadHalf<'_>>) -> Result<TakerToMakerMessage, Error> {
    let len = read_varint(reader).await?;

    let mut buff = Vec::<u8>::with_capacity(len.0 as usize);
    reader.read_exact(&mut buff).await?;

    let message = TakerToMakerMessage::net_deserialize(&buff[..])?;

    Ok(message)
}
```

Along with it a custom `VarInt` reading function (because VarInt's `consensus_decode()` is non-async, so it wont wait until read buffer is filled)
```rust
async fn read_varint(reader: &mut BufReader<ReadHalf<'_>>) -> Result<VarInt, Error> {
    let n = reader.read_u8().await?;

    match n {
        0xFF => {
            let x = reader.read_u64().await?;
            if x < 0x100000000 {
                Err(self::Error::Protocol("Bad VarInt"))
            } else {
                Ok(VarInt(x))
            }
        }
        0xFE => {
            let x = reader.read_u32().await?;
            if x < 0x10000 {
                Err(self::Error::Protocol("Bad VarInt"))
            } else {
                Ok(VarInt(x as u64))
            }
        }
        0xFD => {
            let x = reader.read_u16().await?;
            if x < 0xFD {
                Err(self::Error::Protocol("Bad VarInt"))
            } else {
                Ok(VarInt(x as u64))
            }
        }
        n => Ok(VarInt(n as u64))
    }
}
```
This is the generic scheme of `read` and `send` used by both `maker` and `taker` protocol. Along with `offerbook_sync()` offer downloads.

But so far I am being hit by `UnexpectedEOF` error on both maker and taker side.
taker error 
```shell
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: Serialisation(ConsensusEcode(Io(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" })))buff data: []
', src/offerbook_sync.rs:82:86thread '
tokio-runtime-workernote: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
' panicked at 'called `Result::unwrap()` on an `Err` value: Serialisation(ConsensusEcode(Io(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" })))', src/offerbook_sync.rs:82:86
```
maker error
```shell
error reading from socket: Serialisation(ConsensusEcode(Io(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" })))
```

So far it doesn't seem to be straight forward either. I suppose mainly because we are using async stuffs, and the `read_exact()` method in the `read_message()` function is hitting EOF, when it should just keep on reading.

GeneFerneau commented:
> I tried doing something like that

Do you have a separate branch with those changes? I tried reproducing the errors you list, but every test passes locally with your latest changes.

> I suppose mainly because we are using async stuffs, and the read_exact() method in the read_message() function is hitting EOF, when it should just keep on reading

Not sure what the right answer is here, since a sync function is being called within an async reader. May help to use the `read` function instead. If you point me to a branch with your changes, I can pull them down and help debug.

mojoX911 commented:
> Maybe try s/read_exact/read_line in read_message, similar to current master.

`read_line` won't work because we can't use newline deliminator in byte encoded message. `\n` is just `0x10` in byte, and we can't guarantee that `0x10` won't exists inside the message itself. So trying to use `read_line` will prematurely end the read whenever it finds one `0x10`. 

So the approach was to use a `VarInt` instead of deliminator, so that we can specify upfront how much data to read. 

I have pushed these changes in my branch here https://github.com/mojoX911/teleport-transactions/tree/encoding-trial

Its rough, and I commented out previous read methods, but this should explain what I tried to do. 

Thanks, any suggestion on this would be very helpful.

GeneFerneau commented:
> I have pushed these changes in my branch here https://github.com/mojoX911/teleport-transactions/tree/encoding-trial
>
> Its rough, and I commented out previous read methods, but this should explain what I tried to do.

Thanks for posting your changes.

I pulled them down, and think I've found the bug: you need to set the size of the buffer you pass to `read_exact`.

Something like:

```rust
// either
let mut buff: Vec<u8> = Vec::with_capacity(len.0 as usize);
buff.resize(len.0 as usize, 0);
// or
let mut buff = vec![0; len.0 as usize];
reader.read_exact(&mut buff).await?;
```

Think there is still a hang somewhere, but that fixes the `EOF` issue. Make sure to size the buffer in the maker and taker `read_message`.

mojoX911 commented:
@GeneFerneau Ah thanks a lot.. Yes that seems like the thing I am missing. Will give this a try and report back..

mojoX911 committed:
Add serialization trait

We need our custom serilization called `NetSerilizastion` trait, as
bitcoin's consensus encoded cannot be extended directly to our modified
coinswap structures.

Though this trait heavily dependeds on bitcoin's `consensus_encode`, as
mostly use it for primitive serialization derivation.

mojoX911 committed:
implement `NetSerialization` for messages

The custom serialization trait is implemented for all message variants.

A new `TakerToMakerMessage::TestMessage` message is created for internal
testing message (like killing a maker process from integration tests).

Unit test cases are added for message encoding/decoding roundtrips.

mojoX911 committed:
add dedicated NetSerialization error

mojoX911 committed:
use bitcoin's rand feature

This is required for message encoding unit tests.

mojoX911 committed:
update maker protocol to use byte serialization

Replace the existing json text encoding communications with byte encoded
send and receive functions for maker protocol.

VarInt size demarcation is used to specify length of the message in wire

mojoX911 committed:
update taker protocol with byte serialization

same as maker protocol, use new byte encoding instead of json text.

VarInt is used for message size specification on wire.

mojoX911 committed:
update offer book sync logic

Offer book sync internal logic is modified to use byte encoded messages
instead of json encoding.

mojoX911 committed:
update main

Finally add `serialization.rs` as crate module, and update integration
test's kill sitch.

mojoX911 commented:
I have finally managed to get to the bottom of various byte encoding issue. And it seems now all is working at a satisfactory level. Although I am seeing a little lag in standard coinswap process, that can be something internal to the `read` methods. Will need to investigate more.

So far basic coinswap is working with byte serialization and this PR is ready to have another round of review. I have restructured to commit to make review easier, but its still a lot of code as a whole.

I was thinking about breaking it up into multiple PRs, but that might cause more headache than help. Right now its a complete change that implements and shift all the network level communication to byte encodings.

Some utility method refactoring can be used, but I would like to do that in separate smaller PRs. This one is already big enough in its scope.  

Thanks @GeneFerneau for helping out with the suggestions.

mojoX911 commented:
It seems for some reason codecov compilation is failing and it probably is a grcov upstream issue. Nothing related to this PR.

I am investigating further.

mojoX911 committed:
Wake up workflow
